### ISSUE TITLE
Index projection

### ISSUE THREAD

**Issue by kirillt** [ARK-Builders/arklib-android#38]
It should be easier and more flexible way to work with "favorites" which use index of their parent folder but exploit optimized work with resource collection. Right now, there is awkward `prefix: Path` parameter in methods of `ResourceIndex` interface.

**Comment by kirillt** [comment]
Index projections would also help to simplify `AggregatedIndex` implementation.

The `prefix` parameter looks really weird in case of aggregation: it is useful only with "favorites", but there can't be any "favorite" folder attached to several parent folders. This `prefix` parameter should have it's own interface implementation.

```
class AggregatedIndex: ResourceIndex {
    ...
    override suspend fun allResources(prefix: Path?): Set<Resource> =
        //todo would it help to have Map<ResourceIndex, Path> holding prefixes/root paths?
        //we don't want to do redundant calls to shards which cannot handle `prefix`
        shards.flatMap { it.allResources(prefix) }
            .toSet()
    ...
}
```

**Comment by kirillt** [comment]
Implemented in #45
